C++: Wire up param/arg indirections in data flow

jbj opened the pull request:
This PR is based on #3097, which should be merged first. Only b622d62d3c5250ca6335b28d80a283e8208d4f5c is unique to this PR (for now).

This PR is an attempt to do a principled version of #2737, conservative enough that it applies to data flow. I thought it would just be a matter of wiring up `ReadSideEffectInstruction` to `InitializeParameterInstruction`, but it turns out that the involved operands are imprecise. I hope @dbartol or @rdmarsh2 have perspectives on whether they're supposed to be imprecise and whether the solution in this PR could be improved.

jbj commented:
The test failures look similar but not identical to the undesired changes in https://github.com/Semmle/ql/pull/2704. I'll investigate, but I'd still like to hear comments on the overall approach of this PR.

rdmarsh2 commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -142,28 +142,70 @@ class ExprNode extends InstructionNode {
-class ParameterNode extends InstructionNode {
-  override InitializeParameterInstruction instr;
-
+abstract class ParameterNode extends InstructionNode {
```
I'm a bit nervous about this class becoming abstract, since it could affect customer code that extends it and I've seen that pattern used for RemoteFlowSource in other languages.

jbj replied:
Good point -- we don't want to expose abstract classes.

jbj replied:
I've pushed a commit to avoid `abstract`, make `ParameterIndirectionNode` public, and align the class names with Java's DataFlowUtil.

jbj commented:
I've merged #3137 into this PR, hoping it'll fix the test failures.

rdmarsh2 commented:
re: the operands being imprecise, that was an attempt to represent the uncertainty about what the callee would actually read. I don't think there's a problem with making them precise.

jbj commented:
I've started https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/993/

jbj commented:
The CPP-Differences job showed many new results on openjdk/jdk. Thanks to #3189 I've been able to see that they were FPs, and I was able to find the cause. I'm guessing #2704 is affected by the same problem.

In code like

```
mystruct s;
s.x = tainted();
// s := chi(total = Uninitialized[s], partial = tainted())
sink(s.y);
```

there is (and should be) _no_ taint into `sink` even though there is taint into the `chi` instruction. That's because the address through which `y` is loaded is a fresh `VariableAddressInstruction` rather than anything derived from the chi instruction, and the load is inexact. But if the last line of the example happens in a separate function, then this PR makes it possible (and it's the intended behaviour) for the taint in the chi instruction to end up in the `this` pointer, and from there it can then spread through a `FieldAddressInstruction` (which is a `UnaryInstruction`) to the address of a `LoadInstruction`. It's all due to the interaction between these three rules:

https://github.com/Semmle/ql/blob/e5d3286ee9dfafa07db488d0b1590d89722f4073/cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll#L187-L193

jbj commented:
I've started https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1014/

jbj commented:
I looked through the CPP-Differences. We lost 21 results, and they all look like false positives to me. They were due to field conflation.

I've marked this PR as ready for review. It started out as a PR to add _more flow_, but in practice it now causes _less taint_. The extra flow we're adding was (mostly) already present as taint due to #2737. With the latest change to block `FieldAddressInstruction` we now have less taint than before. This PR should make a difference for field flow, though, since field flow is carried by data flow but not by taint. I hope that merging this PR will also let us remove the workaround from #2737 soon.

jbj commented:
I think this PR is essentially ready for merge, but it doesn't provide much value on its own, so we may not even want it in 1.24 unless it's needed for https://github.com/Semmle/ql/pull/3118.

The FPs removed with this PR might also get removed by https://github.com/Semmle/ql/pull/3219. We're still waiting for the CPP-Differences on that one.

MathiasVP commented:
> I think this PR is essentially ready for merge, but it doesn't provide much value on its own, so we may not even want it in 1.24 unless it's needed for #3118.

I don't think we'll need it for #3118.

jbj commented:
Now that IR field flow has been merged, this PR makes a difference. There are three new results (plus two duplicates) in `fields/ir-flow.expected`. Most other result changes are cosmetic.

I've started https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1114/ to evaluate performance and correctness.

A test is failing in the internal repo because it has some cosmetic differences in the path graph. I think we'll want to revert parts of #2737 at some point to avoid duplicate paths, but it seems to happen only rarely.

jbj commented:
I'll be investigating [the CPP-Differences result changes](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1114//artifact/changes.html). For the record, the summary is:

```
=== Changes per project (+new -fixed) ===

g/git/git                                                                   +9 -0
g/php/php-src                                                               +4 -0

=== Changes per query (+new -fixed) ===

cpp/path-injection                                                          +6 -0
cpp/tainted-format-string                                                   +1 -0
cpp/uncontrolled-allocation-size                                            +6 -0
```

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -438,6 +477,24 @@ private predicate simpleInstructionLocalFlowStep(Instruction iFrom, Instruction
+  // `LoadInstruction` together directly, this rule will break if there's any
+  // reassignment of the parameter indirection, including a conditional one that
+  // leads to a phi node.
+  iTo.(LoadInstruction).getSourceValueOperand().getAnyDef() =
+    iFrom.(InitializeIndirectionInstruction)
```
Note to self: this is the rule that's causing object-to-field flow.

jbj commented:
I went through all the new CPP-Differences result and found that all were FPs caused by field conflation except two results for `cpp/path-injection` in php:
- zend_stream.c:174: new TP
- main.c:2564: new TP

I've pushed a fix.

jbj commented:
I've started https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1124/

jbj commented:
The CPP-Differences shows two FPs that seem to be caused by field conflation. I'll investigate.

jbj commented:
I'm quite confident that the two new FPs are caused by an existing field conflation that's just been amplified by this PR because there's more flow overall. I've opened #3475 with a test case that's independent of this PR.

With that said, I think this PR is still blocked on fixing that field conflation because it introduces two FPs on git/git.

jbj commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -442,6 +488,31 @@ private predicate simpleInstructionLocalFlowStep(Instruction iFrom, Instruction
+  // A read side effect is almost never exact since we don't know exactly how
+  // much memory the callee will read.
+  iTo.(ReadSideEffectInstruction).getSideEffectOperand().getAnyDef() = iFrom and
+  not iFrom.isResultConflated()
```
I don't like this rule. It sends flow into a `ReadSideEffectInstruction`, which actually means there is flow to the result value of that instruction. But it's an instruction without a result!

The rule is here because we need a node to be the `ArgumentNode` in `DataFlowPrivate.qll`, and I've chosen the node for the `ReadSideEffectInstruction`. The alternatives I can see are:
1. Omit this rule and use the most recent definition of the indirection (here named `iFrom`) as the `ArgumentNode`. Then a node could be the argument for multiple calls, leading to confusing path explanations.
1. Create a new synthetic node for this purpose.
1. Change the data-flow library so that flow alternates between `Instruction` and `Operand` nodes. The list of good reasons to do this is starting to get long.

jbj replied:
You can see the effect on path explanations in `argvLocal.expected`, where many new nodes with the awkward name `BufferReadSideEffect` are now there.

jbj commented:
I've started https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1168/

jbj commented:
There will be a semantic merge conflict with #3602.

jbj commented:
I think this PR is ready to go.
- I've fixed the semantic merge conflict with #3602.
- [The CPP-Differences job](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1168/) showed three new results, and they all look like TPs to me in the sense that the library does what it should.
- I've opened https://git.semmle.com/Semmle/code/pull/37069 for the internal changes.

I just added 8f702d4b4945665abafbaa070db3becd5942878c, which improves the `toString` output for the new nodes.

MathiasVP commented on cpp/ql/src/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:
```diff
@@ -442,6 +502,31 @@ private predicate simpleInstructionLocalFlowStep(Instruction iFrom, Instruction
+  exists(InitializeIndirectionInstruction init |
+    iFrom = init and
+    iTo.(LoadInstruction).getSourceValueOperand().getAnyDef() = init and
+    // Check that the types match. Otherwise we can get flow from an object to
+    // its fields, which leads to field conflation when there's flow from other
+    // fields to the object elsewhere.
+    init.getParameter().getType().getUnspecifiedType().(DerivedType).getBaseType() =
+      iTo.getResultType().getUnspecifiedType()
```
Could we still get flow from object to field when the type of the field is equal to the parameter type? I'm thinking about a situation with a binary tree like this:

```cpp
struct Tree { Tree *left, *right; };

Tree source();
void sink(Tree*);

void read_left_subtree(Tree* tree) {
  sink(tree->left);
}
...
Tree tree = source();
read_left_subtree(&tree);
```

jbj replied:
Good point. I'll add a test to find out. But even if the answer is that such conflation is possible, does that mean it's undesirable? If we want conflation between array indexes, don't we then also want conflation between entries in a tree or a linked list?

MathiasVP replied:
You're right. It's probably reasonable to have object to field flow in such situations.
